src/mesh/boundary_info.C

```diff
@@ -1240,25 +1247,66 @@ void BoundaryInfo::boundary_ids (const Elem * const elem,
   // Clear out any previous contents
   vec_to_fill.clear();
 
-  // Only level-0 elements store BCs.  If this is not a level-0
-  // element get its level-0 parent and infer the BCs.
+  // In most of the cases only level-0 elements store BCs.
+  // In certain application (such as time-dependent domains), however, children
+  // need to store BSc too. This case is covered with the _children_on_boundary
+  // flag.
   const Elem * searched_elem = elem;
+
+#ifdef LIBMESH_ENABLE_AMR
+
   if (elem->level() != 0)
+  {
+    // If we have children on the boundaries, we need to search for boundary IDs on the
+    // child and its ancestors too if they share the side.
+    if (_children_on_boundary)
     {
-      if (elem->neighbor_ptr(side) == nullptr)
-        searched_elem = elem->top_parent ();
-#ifdef LIBMESH_ENABLE_AMR
-      else
-        while (searched_elem->parent() != nullptr)
-          {
-            const Elem * parent = searched_elem->parent();
-            if (parent->is_child_on_side(parent->which_child_am_i(searched_elem), side) == false)
-              return;
-            searched_elem = parent;
-          }
-#endif
+      for (const auto & pr : as_range(_boundary_side_id.equal_range(searched_elem)))
+              // We check if we already have this ID, we might have inherited it from our ancestors
+              if (pr.second.first == side &&
+                  std::find(vec_to_fill.begin(), vec_to_fill.end(), pr.second.second) ==
+                      vec_to_fill.end())
+                vec_to_fill.push_back(pr.second.second);
+
+      // Loop over ancestors to check if they have boundary ids on the same side
+      while (searched_elem->parent() != nullptr)
+      {
+        const Elem * parent = searched_elem->parent();
+        if (parent->is_child_on_side(parent->which_child_am_i(searched_elem), side) == false)
+          return;
+
+        searched_elem = parent;
+
+        if (_children_on_boundary)
```

Comment:
This is redundant with the if test above.  The redundant for block above could be eliminated, too, by doing `while (searched_elem)` so you get both the element and its ancestors with the same entry.